distribution/tscript.js

```diff
@@ -7785,6 +7785,45 @@ module.createLabel = function(description)
 		};
 };
 
+
+// Create a canvas icon element with automaticly zoomed contents
+// if the website is zoomed to 200% then the actual width of the
+// canvas is twice as large. The draw function does not need to
+// care about this
+// Fields of the #description object:
+// * draw - function with a "canvas" argument that draws the canvas icon
+// * width - canvas width
+// * height - canvas height
+// * parent - DOM object containing the control 
+// * style - optional dictionary of CSS styles
+module.createCanvasIcon = function(description)
+{
+	let style = {"display": "block", "width": description.width+"px", "height": description.height+"px"};
+	if(description.hasOwnProperty("style")) Object.assign(style, description.style);
+	let canvas = module.createElement({
+		"type": "canvas", 
+		"parent": description.parent,
+		"classname": "tgui", 
+		"style": style,
+	});
+	canvas.width = description.width;
+	canvas.height = description.height;
+	
+	// zoom
+	// TODO: programmaticly detect whenever zoom changes
```

Comment:
I don't think browser zoom is something that needs be handled in such detail. It can't even be guaranteed to work the same in each browser. When taking a look at macOs, for example, Safari, Firefox and Chrome do pixel ratio zooming *when using keyboard shortcuts* (Firefox also when using the scroll wheel). However, *when using the trackpad to zoom*, all three use a different method of zooming, which still results in pixelated icons:

![trackpad zooming in Safari](https://user-images.githubusercontent.com/28811733/103924600-8eb09680-5116-11eb-9ab7-061f22d7d198.png)
*This is the ide in Safari, zoomed to its maximum using the trackpad. Doing this doesn't change the `devicePixelRatio` in either browser.*

To really get the icons to scale properly, it would be necessary to use SVG instead of Canvas. This would however require a rewrite of each icon. For now, the current method (static handling of `devicePixelRatio`) should be fine.

Reply:
That' true, browsers (in particular mobile browsers) handle zoom very differently, and often inconsistently. In my experience, having a higher (say, doubled) resolution and letting the browser and the user do whatever zooming they deem appropriate is a good-enough solution, and definitely far better than assuming that one canvas pixel maps to one device pixel. I have done this with large documents with very many pixels (several A4 pages at 300 dpi) running into hitting performance problems, so it should be feasible for icons to have a quite high resolution.

Reply:
The problem is that icons keep the resolution after they are created. Reloading the website while zoomed in would update the icons with the new resolution.

There are the following ways that zooming icons could work:

1. keep small 20x20 px icons at all zoom levels
2. use 40x40 px icons at all zoom levels, the difference to 20x20 px is barely noticeable at 100% zoom level. For larger zooms it looks definitely better than 20x20
3. keep it as it currently is: use `devicePixelRatio`
  a) use the current `devicePixelRatio` whenever an icon is created -- current
  b) use the initial `devicePixelRatio`, when the page was loaded
  c) try to keep icons up to date with the `devicePixelRatio`, whenever it has changed
4. add an option to the configuration to allow users change the icon size
5. use SVG icons
  a) embedded into the website
  b) created at runtime
  c.) ~external files~

What do you think, which is the best option?

I also thought about adding scaling to the canvas, it might look nicer when the website is zoomed in, but it might also reduce performance

Reply:
As already said: I *really* don't consider this to be a major issue (correct me if you think otherwise). Because of that, let's just leave it as it is and go with 3a for now. *In the future*, we could try to switch to SVG, but I think it's **out of scope for this PR**. I could however create an issue as a reminder.

Reply:
I think that a fixed zoom of 200% is good enough. It shouldn't be any problem when the page is not zoomed in, as it usual is. But when it is zoomed in, the zoomed in icons look nicer without the need of updating. Note that with the `devicePixelRatio` uneven zooms like `150%` are blurry, because lines are rendered at subpixel level. Shrinking `200%` icons to those zooms looks almost the same.

Reply:
While I agree that 200% is good enough, you might be misunderstanding `devicePixelRatio`. It is not supposed to be an indicator of the zoom level but rather the scaling factor of screens. Most current laptop screens (mine included) and many external monitors already have a scaling factor of `2` when completely zoomed out:

<img width="743" alt="Chrome Devtools showing devicePixelRatio being 2" src="https://user-images.githubusercontent.com/28811733/104179106-a6846500-540b-11eb-87b9-b4c60f48f8c2.png">

Using `devicePixelRatio` then is rather a matter of adapting to *screens* instead of *zoom levels*.

Reply:
Technically `devicePixelRatio` is the zoom level on the pixel level. This is what I meant with zoom. It determines to how many physical pixels on the screen a length of `1px` in CSS style sheets correspond to. Note: you might have a 4k monitor but a logical screen size of 1080p, websites and other applications work as if your screen had 1080p, only when rendering at the pixel level every pixel unit is scaled by 2. When zooming in a browser it is meant to change the logical screen size of the website relative to the system logical screen size. So zooming 200% in the browser is physically a zoom of 400% (on the pixel level), this physical value is reflected by `devicePixelRatio`.

Reply:
Okay we seem to mean exactly the same thing, but I still don‘t understand why using a static factor of 2 would be better than using `devicePixelRatio`. If you want to give the icons a higher resolution to account for zooming, people who already have a `devicePixelRatio` of 2 because of their screen don‘t get any benefit at all. Instead, just using `devicePixelRatio` would assure that icons look good in the _default_, zoomed out state for _everyone_. Antialiasing will most likely happen anyway, and is highly dependent on the browser engine. Just that is IMO not enough to dismiss `devicePixelRatio`. If you still want to avoid it, you can still [disable it](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/imageSmoothingEnabled).

Reply:
Since this seems to be going nowhere and is not a real issue for me, let's go with you default of 2.